crypto/context.c

```diff
@@ -299,127 +514,114 @@ int ossl_lib_ctx_is_global_default(OSSL_LIB_CTX *ctx)
+        return ctx->decoder_store;
+    case OSSL_LIB_CTX_ENCODER_STORE_INDEX:
+        return ctx->encoder_store;
+    case OSSL_LIB_CTX_STORE_LOADER_STORE_INDEX:
+        return ctx->store_loader_store;
+    case OSSL_LIB_CTX_SELF_TEST_CB_INDEX:
+        return ctx->self_test_cb;
+#endif
 
-    if (dynidx != -1) {
-        if (!CRYPTO_THREAD_read_lock(ctx->index_locks[index]))
-            return NULL;
-        if (!CRYPTO_THREAD_read_lock(ctx->lock)) {
-            CRYPTO_THREAD_unlock(ctx->index_locks[index]);
+    case OSSL_LIB_CTX_RAND_CRNGT_INDEX: {
+
+        /*
+         * rand_crngt must be lazily initialized because it calls into
+         * libctx, so must not be called from context_init, else a deadlock
+         * will occur.
+         *
+         * We use a separate lock because code called by the instantiation
+         * of rand_crngt is liable to try and take the libctx lock.
+         */
+        if (CRYPTO_THREAD_read_lock(ctx->rand_crngt_lock) != 1)
             return NULL;
+
+        if (ctx->rand_crngt == NULL) {
+            CRYPTO_THREAD_unlock(ctx->rand_crngt_lock);
+
+            if (CRYPTO_THREAD_write_lock(ctx->rand_crngt_lock) != 1)
+                return NULL;
+
+            if (ctx->rand_crngt == NULL)
+                ctx->rand_crngt = ossl_rand_crng_ctx_new(ctx);
         }
-        data = CRYPTO_get_ex_data(&ctx->data, dynidx);
-        CRYPTO_THREAD_unlock(ctx->lock);
-        CRYPTO_THREAD_unlock(ctx->index_locks[index]);
-        return data;
-    }
 
-    if (!CRYPTO_THREAD_write_lock(ctx->index_locks[index]))
-        return NULL;
-    if (!CRYPTO_THREAD_write_lock(ctx->lock)) {
-        CRYPTO_THREAD_unlock(ctx->index_locks[index]);
-        return NULL;
-    }
+        p = ctx->rand_crngt;
 
-    dynidx = ctx->dyn_indexes[index];
-    if (dynidx != -1) {
-        data = CRYPTO_get_ex_data(&ctx->data, dynidx);
-        CRYPTO_THREAD_unlock(ctx->lock);
-        CRYPTO_THREAD_unlock(ctx->index_locks[index]);
-        return data;
-    }
+        CRYPTO_THREAD_unlock(ctx->rand_crngt_lock);
 
-    if (!ossl_lib_ctx_init_index(ctx, index, meth)) {
-        CRYPTO_THREAD_unlock(ctx->lock);
-        CRYPTO_THREAD_unlock(ctx->index_locks[index]);
-        return NULL;
+        return p;
     }
 
-    CRYPTO_THREAD_unlock(ctx->lock);
+#ifdef FIPS_MODULE
+    case OSSL_LIB_CTX_THREAD_EVENT_HANDLER_INDEX:
+        return ctx->thread_event_handler;
+#endif
+
+    case OSSL_LIB_CTX_FIPS_PROV_INDEX: {
+        /*
+         * fips is a separate module which may or may not be loaded,
+         * so we have to do this lazily.
+         */
+        if (CRYPTO_THREAD_read_lock(ctx->lock) != 1)
+            return NULL;
+
+        if (ctx->fips_prov == NULL) {
+            CRYPTO_THREAD_unlock(ctx->lock);
+
+            if (CRYPTO_THREAD_write_lock(ctx->lock) != 1)
+                return NULL;
+
+            if (ctx->fips_prov == NULL) {
+                ctx->fips_prov = meth->new_func(ctx);
```

Comment:
If this were wrapped in "ifdef FIPS_MODULE" I'm not sure why we need the method? Why can't we just handle it in the same way as all the others? Note: context.c is compiled *twice*. Once with FIPS_MODULE defined and once without. The version without FIPS_MODULE defined is linked into libcrypto.so. The version with FIPS_MODULE defined is linked into fips.so so we should be able to use symbols from that module.